base/dcp_client_test.go

```diff
@@ -311,13 +311,117 @@ func TestDCPClientMultiFeedConsistency(t *testing.T) {
 				// only vbucket 0 should have rolled back, expect mutation count to be only vbucketZero
 				mutationCount := atomic.LoadUint64(&mutationCount)
 				require.Equal(t, int(vbucketZeroExpected), int(mutationCount))
+				// check the rolled back vBucket has in fact closed the stream after its finished
+				numVBuckets := len(dcpClient.activeVbuckets)
+				require.Equal(t, uint16(0), uint16(numVBuckets))
 			case <-feed3Timeout:
 				t.Errorf("timeout waiting for first one-shot feed to complete")
 			}
 		})
 	}
 }
 
+func TestContinuousDCPRollback(t *testing.T) {
+	SetUpTestLogging(t, LevelDebug, KeyDCP)
+
+	if UnitTestUrlIsWalrus() {
+		t.Skip("This test only works against Couchbase Server")
+	}
+
+	var vbUUID gocbcore.VbUUID = 1234
+
+	bucket := GetTestBucket(t)
+	defer bucket.Close()
+	dataStore := bucket.GetSingleDataStore()
+
+	// create callback
+	mutationCount := uint64(0)
+	vbucketZeroCount := uint64(0)
+	counterCallback := func(event sgbucket.FeedEvent) bool {
+		if bytes.HasPrefix(event.Key, []byte(t.Name())) {
+			atomic.AddUint64(&mutationCount, 1)
+			if event.VbNo == 0 {
+				atomic.AddUint64(&vbucketZeroCount, 1)
+			}
+		}
+		return false
+	}
+
+	feedID := t.Name()
+	gocbv2Bucket, err := AsGocbV2Bucket(bucket.Bucket)
+	require.NoError(t, err)
+
+	collection, err := AsCollection(dataStore)
+	require.NoError(t, err)
+
+	var collectionIDs []uint32
+	if collection.IsSupported(sgbucket.BucketStoreFeatureCollections) {
+		collectionIDs = append(collectionIDs, collection.GetCollectionID())
+	}
+
+	dcpClientOpts := DCPClientOptions{
+		FailOnRollback:    false,
+		OneShot:           false,
+		CollectionIDs:     collectionIDs,
+		CheckpointPrefix:  DefaultMetadataKeys.DCPCheckpointPrefix(t.Name()),
+		MetadataStoreType: DCPMetadataStoreInMemory,
+	}
+
+	// timeout for feed to complete
+	timeout := time.After(20 * time.Second)
+
+	dcpClient, err := NewDCPClient(feedID, counterCallback, dcpClientOpts, gocbv2Bucket)
+	require.NoError(t, err)
+
+	// function to force the rollback of some vBuckets
+	dcpClient.forceRollbackvBucket(vbUUID)
+
+	doneChan, startErr := dcpClient.Start()
+	require.NoError(t, startErr)
+
+	// Add documents
+	const numDocs = 10000
+	updatedBody := map[string]interface{}{"foo": "bar"}
+	for i := 0; i < numDocs; i++ {
+		key := fmt.Sprintf("%s_%d", t.Name(), i)
+		err := dataStore.Set(key, 0, nil, updatedBody)
+		require.NoError(t, err)
+	}
+
+	// wait for a timeout to ensure client streams all mutations over continuous feed
+	select {
+	case <-doneChan:
+		t.Fatalf("continuous client closed early")
+	case <-timeout:
+		mutationCount := atomic.LoadUint64(&mutationCount)
+		require.Equal(t, uint64(10000), mutationCount)
+	}
```

Comment:
The way this test is written it will _always_ take 20s - this is unfortunate - we should be able to actually assert the count in the `counterCallback` and send a signal when it is done.

This is bad for impatient patient and fast computers and jenkins, which can be slow.